sdk/core/azure-core/azure/core/exceptions.py

```diff
@@ -82,13 +83,96 @@ def map_error(status_code, response, error_map):
     error = error_type(response=response)
     raise error
 
+
+class ODataV4Format(object):
+    """Class to describe OData V4 error format.
+
+    http://docs.oasis-open.org/odata/odata-json-format/v4.0/os/odata-json-format-v4.0-os.html#_Toc372793091
+
+    :param dict json_object: A Python dict representing a ODataV4 JSON
+    :ivar str ~.code: Its value is a service-defined error code.
+     This code serves as a sub-status for the HTTP error code specified in the response.
+    :ivar str message: Human-readable, language-dependent representation of the error.
+    :ivar str target: The target of the particular error (for example, the name of the property in error).
+     This field is optional and may be None.
+    :ivar list[ODataV4Format] details: Array of ODataV4Format instances that MUST contain name/value pairs
+     for code and message, and MAY contain a name/value pair for target, as described above.
+    :ivar dict innererror: An object. The contents of this object are service-defined.
+     Usually this object contains information that will help debug the service.
+    """
+    CODE_LABEL = "code"
+    MESSAGE_LABEL = "message"
+    TARGET_LABEL = "target"
+    DETAILS_LABEL = "details"
+    INNERERROR_LABEL = "innererror"
+
+    def __init__(self, json_object):
+        if "error" in json_object:
+            json_object = json_object["error"]
+        cls = self.__class__  # type: Type[ODataV4Format]
+
+        # Required fields, but assume they could be missing still to be robust
+        self.code = json_object.get(cls.CODE_LABEL)  # type: Optional[str]
+        self.message = json_object.get(cls.MESSAGE_LABEL)  # type: Optional[str]
+
+        if not (self.code or self.message):
+            raise ValueError("Impossible to extract code/message from received JSON:\n"+json.dumps(json_object))
+
+        # Optional fields
+        self.target = json_object.get(cls.TARGET_LABEL, None)  # type: Optional[str]
+
+        # details is recursive of this very format
+        self.details = [
+            self.__class__(detail_node)
+            for detail_node in json_object.get(cls.DETAILS_LABEL, [])
+        ]  # type: List[ODataV4Format]
+
+        self.innererror = json_object.get(cls.INNERERROR_LABEL, {})  # type: Dict[str, Any]
+
+    @property
+    def error(self):
+        import warnings
+
+        warnings.warn(
+            "error.error from azure exceptions is deprecated, just simply use 'error' once",
+            DeprecationWarning,
+        )
+        return self
+
+    def __str__(self):
+        return "({}) {}".format(self.code, self.message)
+
+    def message_details(self):
+        """Return a detailled string of the error.
+        """
+        # () -> str
+        error_str = "Code: {}".format(self.code)
+        error_str += "\nMessage: {}".format(self.message)
+        if self.target:
```

Comment:
Nitpick: technically, `self.code` and `self.message` are both optional, yet we don't guard those/print 'None' as the value. But we suppress `self.details`. I assume that this is intentional.

Reply:
If you look at the `__init__`, I actually accept one of both code or message to be None, but not both at the same time. So the possible print for `__str__` are:
```
(code) message
(None) message
(code) None
```

IOW, they are supposed to be required, but I don't want to fail if they are not here, but I don't want to also have an empty ODataV4 (is `{"foo": "bla"}` a correct OdataV4, with just missing code message? :))

TLDR: on purpose